### ISSUE TITLE
Patch doesn't work on Acer Aspire Vero AV15-52 (InsydeH2O REV5.0)

### ISSUE THREAD

**Issue by betayanme** [SmokelessCPUv2/SREP-Community-Patches#2]
Hello, my equipment is Acer Aspire Vero AV15-52. I tried to use these two baked patch files "Levovo Legion" to unlock the BIOS advanced option, but it didn't work. I want to know how to make these patch files. Use UEFITool to find the string character and then find the corresponding GUID? I tried to do this, but unfortunately, it still doesn't work. Later, I tried to use the patch "PREDATOR TRITON 300". After entering the BIOS, I could see that there were several previously hidden options in the Advanced menu, but unfortunately, these options were useless to me (at the same time, the hidden "Advanced" and "Power" tabs still did not appear). So can you tell me in detail how to make a patch? Thank you very much!

**Comment by betayanme** [comment]
The problem has been solved. 
Reason: "SREP" no longer supports the BIOS of Intel 12th generation processors, and must only modify and replace the BIOS file's GUID to unlock the Advanced option.

**Comment by SmokelessCPUv2** [comment]
SREP has no issue on 12 gen, every bios is different so trying to use config backed for another model will fail..
And I don't think that is an issue related to SREP

**Comment by betayanme** [comment]
I'm really sorry, you're right. I may not be able to express myself clearly because of mental problems. I mean, I tried to make a patch configuration file myself, but the SREP patch for Acer laptop with Intel 12th generation processor didn't work (but it was suitable for Intel 11th generation Acer laptop BIOS). This may be due to the modification of the underlying BIOS. Replacing hexadecimal values directly is invalid (at least I tried). So I finally tried to modify the "BIOS.fd" file to unlock the advanced options.

**Comment by jrmoore** [comment]
Did you get it in the end @betayanme?

I can probably write a patch for that Vero, in 12th gen Intel machines from Acer we can still leverage the ODM module. I have no way to tell for sure without a dump though, there's no firmware published for that one in its support page (and navigating the Update catalog is a pain).

**Comment by betayanme** [comment]
> Did you get it in the end @betayanme?
> 
> I can probably write a patch for that Vero, in 12th gen Intel machines from Acer we can still leverage the ODM module. I have no way to tell for sure without a dump though, there's no firmware published for that one in its support page (and navigating the Update catalog is a pain).

Yes, when I was looking for BIOS firmware in Acer official website, I found that only this model had no BIOS to download. Later, I paid someone else to help me write a patch for the BIOS of this model, but it didn't work.......Finally, I used "Intel ME System Tool (Flash Programming Tool)" to extract the BIOS, and then I paid someone again to help me modify the BIOS and unlock the advanced options.

If you need the BIOS I extracted and the modified BIOS, you can download it from the following link: https://drive.google.com/drive/folders/1VJjxvllWbGzdcyjHRyiKNH61wZr_j8jl?usp=sharing

If you want to FLASH this BIOS, please take your own risks.

**Comment by jrmoore** [comment]
> Yes, when I was looking for BIOS firmware in Acer official website, I found that only this model had no BIOS to download.

I think that's because it's one of the newer Veros, at least I haven't seen anyone with it around just yet. I think I would notice because of their appearance, even if they are more muted now the 51s looked different than the usual laptops (I kinda dig that look).

That reminds me, the fact that there is no published firmware for the machine means updates weren't pushed yet, but they will be over time (not that Acer has the best support rate ever, but still) and they can be delivered over Windows Update. If you'd like to keep the modifications, beware of it (if you use Windows that is) and perhaps disable the firmware interface in the device manager, you'd be looking for something like this:

![Firmware upgrade interface](https://user-images.githubusercontent.com/311922/220961555-ba7e967f-666d-4ab5-a873-ab63dba8062e.png)

<hr>

> Later, I paid someone else to help me write a patch for the BIOS of this model, but it didn't work.......Finally, I used "Intel ME System Tool (Flash Programming Tool)" to extract the BIOS, and then I paid someone again to help me modify the BIOS and unlock the advanced options.

Damn... you were determined to get it unlocked! Paying for those modifications, is it how it usually is? I mean, work is work, of course, but.. feels wrong, specially if it doesn't pan out.

In any case yes, that's usually how it goes on Intel machines to flash new firmware without reprogramming the chip directly, removing the (more than likely) BIOS lock and possible FPRR protections beforehand.

> If you need the BIOS I extracted and the modified BIOS, you can download it ...

> If you want to FLASH this BIOS, please take your own risks.

Thanks! I'll take a look at it, an original BIOS dump if enough though, depending on whether I've seen similar structures before or not it may take a while, but I should be able to provide a SREP config for it. I asked in case you were still in need of it being done, I wanted to give it a go (I don't own a Vero).

As for the last part, you're correct, while it is technically possible to flash someone else's dump into a machine of the same line it is ill-advised. BIOS dumps differ from the images that we get out of the upgrade packages in that they also contain data unique to each device like crypto keys or serial numbers, for example. Those with the same machine should ask the original modder to replicate it with a dump of their own, not use one found around (unless it's a party what they're looking for 🤣).

Having SREP, (which, by the way, thank very much @SmokelessCPUv2, I understand it's not maintained at the moment, but I'm grateful nonetheless), being able to test modifications at runtime without needing to flash is amazing. I wonder how it'd go with those machines that have the root of the chain of trust deeply embedded though...

**Comment by jrmoore** [comment]
Okay I'm done, for the firmware version in there we can still patch a module different from `SetupUtility` to unlock access to the (proper) advanced and power tabs, that's not the case for newer Acers though, two low-powered Aspires already have their firmware up and even though the module is present there's no trace of the protocol, but I didn't study those in depth.

The patch I wrote could be more succinct there, but I like the flow this way haha, the other patches on `SetupUtility` is to allow access to the overclocking area should people want to.

Now, I also took a look at the mod they did for you and it's different. These laptops come with an Advanced tab in the setup utility that serves more or less to handle VT-x and VT-d (there seems to be something else there now too), the person that modded your BIOS dump replaced the form ID for that tab with the one for the other Advanced tab in `SetupUtilityApp` so when you go to it the later is shown instead of the former.

It is functional, but incomplete in my humble opinion, I like to have full range of options, if at all possible that is. I also wrote the equivalent SREP operations for that modification, in case someone prefers it. I tagged the archive with the firmware version in case they change things and needs adjustments going forward, but for now at least it should work.

[AV15-52 configs, firmware v1.04.zip](https://github.com/SmokelessCPUv2/SREP-Community-Patches/files/10816497/AV15-52.configs.firmware.v1.04.zip)

**Comment by betayanme** [comment]
>That reminds me, the fact that there is no published firmware for the machine means updates weren't pushed yet, but they will be over time (not that Acer has the best support rate ever, but still) and they can be delivered over Windows Update. If you'd like to keep the modifications, beware of it (if you use Windows that is) and perhaps disable the firmware interface in the device manager, you'd be looking for something like this:
Thank you for reminding me. I have already stopped Windows Update and suspended it until 2999:
https://drive.google.com/drive/folders/1wHI-CgwUGvkh1d5Nmzy6a5OvPOvO9NSO?usp=sharing

>In any case yes, that's usually how it goes on Intel machines to flash new firmware without reprogramming the chip directly, removing the (more than likely) BIOS lock and possible FPRR protections beforehand.
Yes, so before I extract the BIOS, I will enter the following command in “grub” to disable the BIOS lock: 
setup_var PchSetup 73d 00
setup_var PchSetup 1c 00

>It is functional, but incomplete in my humble opinion, I like to have full range of options, if at all possible that is. I also wrote the equivalent SREP operations for that modification, in case someone prefers it. I tagged the archive with the firmware version in case they change things and needs adjustments going forward, but for now at least it should work.
Thank you very much. I tested it. The patch you made worked. It's perfect.

**Comment by betayanme** [comment]
>That reminds me, the fact that there is no published firmware for the machine means updates weren't pushed yet, but they will be over time (not that Acer has the best support rate ever, but still) and they can be delivered over Windows Update. If you'd like to keep the modifications, beware of it (if you use Windows that is) and perhaps disable the firmware interface in the device manager, you'd be looking for something like this:
>
Thank you for reminding me. I have already stopped Windows Update and suspended it until 2999:
https://drive.google.com/drive/folders/1wHI-CgwUGvkh1d5Nmzy6a5OvPOvO9NSO?usp=sharing

>In any case yes, that's usually how it goes on Intel machines to flash new firmware without reprogramming the chip directly, removing the (more than likely) BIOS lock and possible FPRR protections beforehand.
>
Yes, so before I extract the BIOS, I will enter the following command in “grub” to disable the BIOS lock: 
setup_var PchSetup 73d 00
setup_var PchSetup 1c 00

>It is functional, but incomplete in my humble opinion, I like to have full range of options, if at all possible that is. I also wrote the equivalent SREP operations for that modification, in case someone prefers it. I tagged the archive with the firmware version in case they change things and needs adjustments going forward, but for now at least it should work.
>
Thank you very much. I tested it. The patch you made worked. It's perfect.

**Comment by jrmoore** [comment]
> Yes, so before I extract the BIOS, I will enter the following command in “grub” to disable the BIOS lock:
> setup_var PchSetup 73d 00
> setup_var PchSetup 1c 00

Yep, that's a way of doing it, and one of the only ways for machines with Insyde based firmware that have newer Intel processors (that I know of, another would be throuhg RU.efi). There used to be a (leaked) Windows utility to edit those variables, but it's not usable anymore.

Those offsets, 0x1C (BIOS lock) and 0x73D (FPRR) are model dependent, both are in `PchSetup` (sometimes in a custom one too), I check *where* they are in the `SetupUtility` module of the firmware first to know for sure, but from what I'm reading, those locations may be linked to the Intel ME version the system has.

Anyway, I digress haha, I'm glad it works, it can be added to list 🙂.

**Comment by betayanme** [comment]
One more thing, I'd like to ask how you made the "SREP-Patch"？I would appreciate it if you could tell me.

**Comment by betayanme** [comment]
> > Yes, so before I extract the BIOS, I will enter the following command in “grub” to disable the BIOS lock:
> > setup_var PchSetup 73d 00
> > setup_var PchSetup 1c 00
> 
> Yep, that's a way of doing it, and one of the only ways for machines with Insyde based firmware that have newer Intel processors (that I know of, another would be throuhg RU.efi). There used to be a (leaked) Windows utility to edit those variables, but it's not usable anymore.
> 
> Those offsets, 0x1C (BIOS lock) and 0x73D (FPRR) are model dependent, both are in `PchSetup` (sometimes in a custom one too), I check _where_ they are in the `SetupUtility` module of the firmware first to know for sure, but from what I'm reading, those locations may be linked to the Intel ME version the system has.
> 
> Anyway, I digress haha, I'm glad it works, it can be added to list 🙂.

One more thing, I'd like to ask how you made the "SREP-Patch"？I would appreciate it if you could tell me.

**Comment by jrmoore** [comment]
> I'd like to ask how you made the "SREP-Patch"?

That question is easy and hard at the same time haha, easy because SREP is well documented about what you can do through it in its repo. For example, you can get to a module that has already been loaded, you can load one that hasn't been loaded yet, and you can patch them at a specific offset, or instruct it to look for a pattern of consecutive bytes and replace it for others.

Difficult because getting to the point where you know what to replace and with what requires engineering the firmware of the device. At least it can be done with a firmware upgrade that a manufacturer provides, a BIOS dump is not strictly required for approaching things through SREP, it only was for your machine because there was none available.

There are so many variables that sometimes it's a nightmare, I've seen modifications that I don't even comprehend how it is they got there, even with the code in front of my eyes. Insyde, Phoenix, AMI, ... different platforms = different approaches, some are encrypted, others are protected and even changes to firmware would be rejected. They all adhere (better or worse) to the same UEFI specification (plenty to read there on its own), but manufacturers not only use those firmware platforms, they also customize them. A Lenovo based on Insyde is different than one from Acer, for example.

That's why there aren't (to my knowledge) generic guides on how to mod firmware, there are several written in different sites about various topics, many of them useful, blog posts of people who already did some research into a specific machine and many resources in languages we may or may not be able to read and understand (Russian, Chinese, etc.).

<hr>

Let's go over the modification that guy did for you as an example, because it doesn't involve disassembling the code of the modules to figure out where it is or how they do to hide or not show tabs, or how to somehow exploit or bypass obfuscation.

We start with a BIOS dump in this case, of an Intel-based system from Acer that uses InsydeH2O Revision 5. Forgetting about the business of disabling the locks (which we'll get to see as well) first we need to do is unpack the `SetupUtility` module, in Insyde that's where the code for the grey and blue configuration screen lies.

There are several ways to do it, one of them would be using [UEFITool](https://github.com/LongSoft/UEFITool), so we open the app and drop in the dump:

![1](https://user-images.githubusercontent.com/311922/221190960-82d72667-9f6d-4734-b29d-57d050ba5fa2.png)

I'll try not to take detours of this will be a novel, so we'll search for the module directly (Ctrl+F), it may be have the usual GUID, but just in case we'll search by text:

![2](https://user-images.githubusercontent.com/311922/221191478-e0ed6c2f-9cff-42fb-888a-faa6ab06d1c4.png)

Double clicking one of the references takes us there, and we want to right click over the PE32 image and extract its body to a folder. Now that file has many definitions and overall structure of the setup routine, it is a valid UEFI module we can drop in a disassembler (IDA, Ghidra, ..) to try to make sense of the code, but we won't be needing it for this. Since we're still in UEFITool we're going to extract another module as well, `SetupUtilityApp`, that's where we do the swap.

Now, there's another program, [Universal IFR Extractor](https://github.com/LongSoft/Universal-IFR-Extractor), that we'll need to get something "readable" out of the `SetupUtility` module, opening that app and feeding it the extracted module gets us a text file with descriptions, offsets, values, of the setup structure.

Opening it we begin to see stuff, I have annotated some things in there:

![4](https://user-images.githubusercontent.com/311922/221195317-b7b2de09-1e23-40a2-a9fa-ed082dd77419.png)

OK, so what to do next? Get the GUIDs of both Advanced tabs, so we search for either its name until we get near there, or its offset + 4, which is where IFR Extractor starts documenting the form:

![5](https://user-images.githubusercontent.com/311922/221197345-b9bb189f-0efa-4e06-a564-5a6784731306.png)

![6](https://user-images.githubusercontent.com/311922/221197914-11f262e1-32b2-4824-ab03-b84c8239ea9c.png)

So hey, we found the two Advanced tabs and their identifiers, how do we make make the setup routine display one instead of the other? That's somewhere else, in the app itself, that other `SetupUtilityApp` we extracted earlier as well.

Since we know the second, less interesting, Advanced tab is shown by default its ID ought to be referenced there, so lets open that file in an hexadecimal editor. No need for anything fancy, just like Windows' Notepad was enough for before HxD is more than enough now:

You'll see in that picture that I'm not looking for the whole ID, that's because sometimes things are encoded differently, after all they try to make things not as easy to break (we welcome modifications, but in a tightly controlled environment.. not so much). Instead I'm going to look for the last part of it and see if it appears, it will in this case, but if it didn't one could try reversing the bytes perhaps, or going into a disassembler and getting dirty with the code.

![7](https://user-images.githubusercontent.com/311922/221202685-c8f0f244-4a7a-4111-98ca-99d4c76b3cfc.png)

![8](https://user-images.githubusercontent.com/311922/221201946-56b6e9b9-c013-47c4-b6b9-b3cede6fcf21.png)

Okay, so we're at a point where we can write a patch for SREP, I'm sure you're still wondering *"WTF, how do I know it's there? I mean, sure, I can rule out the first search result because it doesn't fully match the GUID I was looking for, but what gives?"* And in this case it's because I know that's the lookup table for the form IDs that are shown by default, because not only I looked at the IDs of the tabs of interest, but also those that are shown by default and they're all there:

![9](https://user-images.githubusercontent.com/311922/221204946-81a63a74-eac7-4b9b-aead-b5a0ccc413fb.png)

<hr>

So writing an SREP config file for this is now the easy part:
 * First we need to select the module, we know we want to patch `SetupUtilityApp` to switch one GUID with the other.
 * Since that's an app (the actual GUI for the setup), it's not loaded by the time SREP is so we need to load it from the firmware first: 

```
Op LoadFromFV
SetupUtilityApp
```

* We know the offset where the GUID starts, we could go that way, but since there are a lot of bytes to replace and they only appear once in there it's better to go with patterns. Patterns are search & replace done in bytes, so hexadecimal, and they're to be done in the order they appear in the file:

```
Op Patch
Pattern
ED9786EE4C90D640A4D2CC0762DDFC76
```

* That would be what we're looking for, the form ID of the basic Advanced tab, but remember that the first half has its bytes swapped, so instead of `C6D4769E-7F48-4D2A-98E9-87ADCCF35CCC` we need to do use `9E76D4C6-487F-2A4D-98E9-87ADCCF35CCC`. It's easy to see in the middle for example, `4D2A` becomes `2A4D`:

```
9E76D4C6487F2A4D98E987ADCCF35CCC
```

* Finally, we just need to launch the setup utility and we're done:

```
Op Exec
End
```

<hr>

That's pretty much how the modification the guy did would go, (s)he may have approached it differently, but overall that's how I'd have gone about it. The text representation that comes out of IFR Extractor is more useful too, for example, there you can see where the BIOS lock and FPRR bytes are for that machine:

![10](https://user-images.githubusercontent.com/311922/221209035-f6efa36f-2d84-465d-be9d-a5c75377473e.png)

![11](https://user-images.githubusercontent.com/311922/221209067-c5b39916-3555-4367-87b4-81ebbb144c20.png)

![12](https://user-images.githubusercontent.com/311922/221209115-3284d4e7-a41d-48db-92d4-d85705553c4b.png)

And it also shows areas that would require additional changes to enable access to them, for example, just swapping the tabs you wouldn't be able to access the overclocking menu, it'd be grayed out (I took it into account on the patch I wrote, it's those two patches applied to the `SetupUtility` module).

So, overall... I wouldn't say it's *easy*, because one needs to know about some things to get there and do that swap, but it's a easy patch to write, it may take some time to get to see these things but it's not unsurmountable if one is determined. If we needed to fire up a diassembler, go through the code (it's x86-64 assembly) understand its flow, what each of the subroutines do, maybe even try to bring it to some C-like pseudocode to try to get some clarity... yeah, that's way more complicated.

Other BIOS platforms work differently though, I believe AMI's Aptio 4 has some utility to directly configure the BIOS and for Aptio 5 a guy made a webapp that takes as inputs several files (some extracted, like through UEFITool, others from some other utilities) and shows you also and advanced configuration thing; but I haven't delved into that yet, there's still much to learn.
